File: LeetCode/3.Longest_Substring_Without_Repeating_Characters/Mario/3.Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();

        int n = s.length();
        char[] chars = s.toCharArray();
        int i = 0, j = 0;
        int res = 0;
        while (j < n) {
            char ch = chars[j];
            if (set.contains(ch)) {
                set.remove(chars[i]);
                i++;
            } else {
                set.add(ch);
                j++;
                res = Math.max(res, j - i);
            }
        }
        return res;
    }
}

File: LeetCode/3.Longest_Substring_Without_Repeating_Characters/Mario/3.Longest_Substring_Without_Repeating_Characters.py
#!/usr/bin/env python3


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = {}  # key: char, value: index
        left = 0
        res = 0

        for right, char in enumerate(s):

            if char in dic and dic[char] >= left:
                left = dic[char] + 1

            dic[char] = right
            res = max(res, right - left + 1)

        return res
